src/Symfony/Component/Messenger/Middleware/LockMiddleware.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Middleware;
+
+use Symfony\Component\Lock\LockFactory;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Stamp\LockStamp;
+use Symfony\Component\Messenger\Stamp\ReceivedStamp;
+
+final class LockMiddleware implements MiddlewareInterface
+{
+    public function __construct(private LockFactory $lockFactory)
+    {
+    }
+
+    public function handle(Envelope $envelope, StackInterface $stack): Envelope
+    {
+        $stamp = $envelope->last(LockStamp::class);
+        if (!$stamp instanceof LockStamp) {
+            return $stack->next()->handle($envelope, $stack);
+        }
+
+        if (null === $envelope->last(ReceivedStamp::class)) {
+            $lock = $this->lockFactory->createLockFromKey($stamp->getKey(), $stamp->getTtl(), autoRelease: false);
+            if (!$lock->acquire()) {
```

Comment:
Why no blocking here? Will it simply be marked for retry? If so, maybe it worth a comment here, since the execution flow is split into many middlewares, it might worth the shot explaining why you do this.

Reply:
A blocking acquire have a lot of drawback since it's slower.
If you're generating a lot of messenger message, your middleware won't follow.
I already got CPU issues in production because of blocking acquire.

But we could introduce a `shouldBeBlocking` parameter to the Stamp if needed.

Reply:
Yes that's what I was thinking but it doesn't answer the second question !

Reply:
Rereading the implementation and the param `$blocking`, I don't think there is anything to change/add.

The purpose of the LockMiddleware is to avoid having duplicated similar message in the Queue. There is no reason to wait for the first message to be treated before adding the second message ; that already the purpose of a queue! So the only interesting behavior here is to drop the second message when the first is not treated yet.

Reply:
Alright! - I was thinking about use cases where you'd tag messages that for example would create lock contention on the database using the same key: you may want to ensuring that all would pass one after another but all messages must be consumed.*

I happened to implement such locking mechanism using a similar middleware on client projects because of this. In this use case, all messages are valid, you just don't want them pass simultaneously to avoid too many transactions failures and retries.

Reply:
> you just don't want them pass simultaneously

Isnt the principle of a queue to consume them one by one from the queue ?

Reply:
@pounard, I had a similar problem and was hoping that LockMiddleware could solve it.

@VincentLanglet, In case all messages have to be processed strictly in order, simply dropping the message for re-processing will create a mess in the queue even if we manually (in the handler) check if the message should be processed and resend it if necessary.

It would also be nice to be able to customize the strategy of handling repeated messages (resend/reset)

Reply:
@VincentLanglet,

> Isnt the principle of a queue to consume them one by one from the queue ?

A simple example: you are using RabbitMQ and you have multiple handlers - whether you like it or not, messages will be sent for processing one by one, but will be processed in parallel (because you have multiple handlers)

Of course, you can use only one handler, but in that case you will have to create a separate queue for such messages, or use some specific transport settings, but it is not always reasonable.

Reply:
A lock middleware is definetly not the solution for this need. It would block a whole process while waiting for dispatch the new message.

A blocking lock is not a good idea

Reply:
> Isnt the principle of a queue to consume them one by one from the queue ?

A bus is not always simply a queue. We always spawn 2 to 4 parallel workers per default when we run a project based upon messenger, or any other message bus.

Reply:
> A lock middleware is definetly not the solution for this need. It would block a whole process while waiting for dispatch the new message.

In our own custom implementation, we didn't always block: first iterations we did, we fallbacked at some point to resend the message in queue for later consumption. In this scenario, messages were pushed back as if they were just sent, letting other messages that arrived in the meantime pass before them. It worked fine in most cases.

Hence my question about what happens in the lock fails? Is the message sent back for retry?

Reply:
> In our own custom implementation, we didn't always block: in first iterations at least, then we fallbacked at some point in a different algo where we resent the message in queue for later consumption. In this scenario, messages were pushed back as if they were just sent, letting other messages that arrived in the meantime pass before them. It worked fine in most cases.

I'm sorry but I don't understand this dicussion then, since you talk about another feature and another implementation. We both agree the LockMiddleware won't work for your custom implementation, but that's definetly not the purpose.

This middleware is about deduplication and dropping the duplicated message. So if the lock cannot be aquire, the message is dropped. There is no retry queue.

Reply:
Yes, but the way you implemented it using a stamp with a key is exactly the way I'd implement my own need.

I guess it would worth it expanding the feature to:

1. allow it to be blocking using a boolean optional parameter on the stamp (won't use it myself, but why not)
2. allow messages to be requeued instead of being dropped using another boolean optional parameter.

With those two minor additions to the stamp, and a single middleware, all use cases should be covered.

Your PR is a very good idea, and I wouldn't want to have a second middleware whereas this one exists and does exactly what I need except for a single line of code.

Reply:
> I guess it would worth it expanding the feature to:

This PR is already opened for 6 month without any changes, and still waiting review.

Let's work with small step.
First merging this.
Then, in another PR adding the new feature you need.

Better is the enemy of good.

Reply:
> Better is the enemy of good.

Agreed. I was thinking that maybe, with a wider range of valid use cases, it might become more interesting to reviewers.

But as you wish. If your PR passes as-is, I'll open new ones when I will experience trouble with locking :smile: 

Thank you for answering.

Reply:
Wait, this is major concern, I don't think it's resolved at all.
The minimum would be to make the behavior crystal clear: if you add the stamp, messages will be dropped.
Given the discussion, I wonder if the name of the middleware isn't the confusing part.
A lock waits or delays, that's the common expectation.

Reply:
> A lock waits or delays, that's the common expectation.

Why would it be the common expactation ?

In the symfony documentation https://symfony.com/doc/current/components/lock.html there is a lot of example with  the following code
```
$lock = $factory->createLock('pdf-creation');
if (!$lock->acquire()) {
    return;
}
```
and the blocking parameter is false by default.

I would say that a lock waits or skip the task, with a defaut behavior of skipping the task.

Reply:
The very fact that we have this discussion means there's a topic to clarify. Please mind the rest of my comment.

Reply:
> The very fact that we have this discussion means there's a topic to clarify. Please mind the rest of my comment.

Then, what do you prefer
- A blocking parameter on the lockstamp ?
- Renaming the lockstamp and the lockmiddleware (to what ?)

To me this is very related to the lock behavior since it require the lock component and uses lock key.

Reply:
Like this https://github.com/symfony/symfony/pull/54141/commits/72aae9446f52176f74c55f3a405614ba010b864d ?
Not sure

Reply:
LGTM, with an `@param` :)

WDYT @pounard?

Reply:
@nicolas-grekas, @VincentLanglet

Note that this implementation blocks the main thread in which the message was sent, so you have to wait, even if you are using asynchronous transport, before you can send the message again. It seems to be correct to block the thread (if necessary) after the message is received and then take some action

Reply:
That's exactly my point:
- Blocking this lock is not useful in the main thread
- I can add another lock at the reception of the message like https://github.com/symfony/symfony/pull/54141/commits/d26a2c5a5ac2041d0e53ca093f8e3495b32b68ab but we're starting to add lot of feature for different use cases and people might keep continuing asking new mode with usecase delaying this PR. It would be easier to add modes/beahvior one by one.

Anyway would something like https://github.com/symfony/symfony/pull/54141/commits/d26a2c5a5ac2041d0e53ca093f8e3495b32b68ab satisfy you @lermontex ?

Reply:
@VincentLanglet, That's better, at least now it's clear what will happen to the message. Thank you!

Reply:
> Adding a required argument so that people are required to be explicit?

I'm totally fine with this, we may add a non-blocking-and-requeue behavior later if necessary.

Yet in the use case I had in the past, the non-blocking-and-requeue behavior is the one I needed the most, but if we doesn't reach a consensus about it, then I'm OK to drop it and attempt a new PR in the future.

Reply:
> Anyway would something like [41c0b2f](https://github.com/symfony/symfony/pull/54141/commits/41c0b2f10aa92ee03bcd2a82ddb6e7a324409f98)satisfy you @lermontex ?

It seems unnecessary complexity to me, I'd wish to have simply three mode:
- Lock and block.
- Lock and discard on lock failure.
- Lock and re-queue message on lock failure.

All at consumption, not upon dispatch, a dispatched message doesn't guarantee parallel consumption.

With those three modes, I think it should cover all use case:
- Two similar messages received at the same time drops the duplicated ones (lock or discard).
- Two messages than cannot be treated at the same time can block and wait (lock and block).
- Two messages than cannot be treated at the same time under pressure and you don't want a bus thread to be blocked can continue its operation while the message is re-queued (lock and re-queue).

Third scenario seems the most complicated one but the most effective performance wise for the scenario where you have messages that cannot be treated concurrently: treat the lock failure as an error and let the retry algorithm handle it.

The first scenario doesn't completely fix what you meant to fix in the first place though: in the case were you have duplicated messages, it would prevent them from being processed all at once, but it will not prevent them from being processed one after another if somewhat, they're being sent at the same time, but treated at different times because of some other messages dispatched at the very same moment but delaying the processing between the two.

Reply:
> > Anyway would something like [41c0b2f](https://github.com/symfony/symfony/pull/54141/commits/41c0b2f10aa92ee03bcd2a82ddb6e7a324409f98)satisfy you @lermontex ?
> 
> It seems unnecessary complexity to me, I'd wish to have simply three mode:
> 
> * Lock and block.
> * Lock and discard on lock failure.
> * Lock and re-queue message on lock failure.
> 
> All at consumption, not upon dispatch, a dispatched message doesn't guarantee parallel consumption.
> 
> With those three modes, I think it should cover all use case:

No it doesn't. Especially because you're focused on your use case.

I personally don't want to lock on consumption. I want to lock on dispatch.
I have process which dispatch lot of message often duplicate. Process is quick, but I don't need to plan processing 10 times the message.  When the first message is not processed the second one will give the same result, so is unneeded, then I don't want to add it to the queue.

I assume you didn't read the initial issue then https://github.com/symfony/symfony/issues/54126 it's talking about Deduplicating message IN QUEUE.

So I want a lock and discard message in queue, when YOU want a lock a discard on process ; 2 different use case.
People might also prefer to release the lock before consuming the message or after the consumption ; I use the 2 differents use case. So now we have 3 use case.

I copy this for lock and block to be explicit. I personnaly don't care bout this feature.

For the lock and retry, please open another PR. I won't list and implement every possible use case.

Reply:
@VincentLanglet

Ohhhhh OK my bad. Now I reread more thoroughly the original issue, I do better understand it. I do sincerely apologize for this, I was mistaken.

So I guess your original PR before I got here was the right one considering the issue topic. Now, can I make a suggestion? As your issue states, the initial goal is to **deduplicate** messages **in queue**.

Why wouldn't you simply rename the middleware to a less generic and conflicting name, such as **DuplicateMiddleware**?

I'm sorry for hijacking, but now I get that the name mislead me all along. Rename it, make it naive-simple just as you did in the first place, and explain better in code documentation what it really does and what is its purpose.

Reply:
Definetly a good idea ; let's go with one stamp/middleware by usecase. :)